test/unit/test_domain_decomposition.cpp

```diff
@@ -575,3 +605,65 @@ TEST(domain_decomposition, invalid) {
         EXPECT_THROW(domain_decomposition(rec, ctx, groups), invalid_gj_cell_group);
     }
 }
+
+struct gj_symmetric: public recipe {
+    gj_symmetric(unsigned num_ranks, bool fully_connected):
+        ncopies_(num_ranks),
+        fully_connected_(fully_connected) {}
+
+    cell_size_type num_cells_per_rank() const { return size_; }
+    cell_size_type num_cells() const override { return size_*ncopies_; }
+    arb::util::unique_any get_cell_description(cell_gid_type) const override { return {}; }
+    cell_kind get_cell_kind(cell_gid_type gid) const override { return cell_kind::cable; }
+
+    std::vector<gap_junction_connection> gap_junctions_on(cell_gid_type gid) const override {
+        unsigned shift = (gid/size_)*size_;
+        switch (gid % size_) {
+            case 1 :  {
+                if (!fully_connected_) return {};
+                return {gap_junction_connection({7 + shift, "gj"}, {"gj"}, 0.1)};
+            }
+            case 2 :  {
+                if (!fully_connected_) return {};
+                return {
+                    gap_junction_connection({6 + shift, "gj"}, {"gj"}, 0.1),
+                    gap_junction_connection({9 + shift, "gj"}, {"gj"}, 0.1)
+                };
+            }
+            case 6 :  return {
+                gap_junction_connection({2 + shift, "gj"}, {"gj"}, 0.1),
+                gap_junction_connection({7 + shift, "gj"}, {"gj"}, 0.1)
+            };
+            case 7 :  {
+                if (!fully_connected_)  {
+                    return {gap_junction_connection({1 + shift, "gj"}, {"gj"}, 0.1)};
+                }
+                return {
+                    gap_junction_connection({6 + shift, "gj"}, {"gj"}, 0.1),
+                    gap_junction_connection({1 + shift, "gj"}, {"gj"}, 0.1)
+                };
+            }
+            case 9 :  return { gap_junction_connection({2 + shift, "gj"}, {"gj"}, 0.1)};
+            default : return {};
+        }
+    }
+
+    cell_size_type size_ = 10;
+    unsigned ncopies_;
+    bool fully_connected_;
+};
+
+TEST(domain_decomposition, symmetric_groups) {
+    auto ctx = make_context();
+    for (int nranks = 1; nranks < 20; ++nranks) {
+        for (int rank = 0; rank < nranks; ++rank) {
+            ctx->distributed = std::make_shared<distributed_context>(dummy_context{rank, nranks});
+            for (const auto& R: {gj_symmetric(nranks, true), gj_symmetric(nranks, false)}) {
+                // NOTE: This is a bit silly, but allows us to test _most_ of
+                // the invariants without proper MPI support. If we could get `gather_gids` to
```

Comment:
just curious: can you think of a way to make `gather_gids` work?

Reply:
Not really without major surgery. We could split this function into three, tentatively:
1. register groups, called in parallel and shoving the data into the decomposition
2. gather_gids, collective call
3. actually construct the decomposition

but that'll require putting the data threaded between those calls into the decomposition
since they must be independent and pass around some state.